ports/esp32/README.md

```diff
@@ -26,59 +26,122 @@ There are two main components that are needed to build the firmware:
 ```
 
-It is recommended to use a Python virtual environment if your system package
-manager already provides these libraries, especially as the IDF v4.x is
-currently incompatible with pyparsing 2.4 and higher.
+Note: If you get an error about `--no-site-packages`, then modify
+`esp-idf/tools/idf_tools.py` and make the same change as [this
+commit](https://github.com/espressif/esp-idf/commit/7a18f02acd7005f7c56e62175a8d1968a1a9019d).
+
+2. Downloading pre-built toolchain manually (IDF 3.x and 4.x)
+=============================================================
+
+Note: while this works with 4.x, if you're using the 4.x IDF, it's much
+simpler to use the guide above, which will also get a more recent version of
+the toolchain.
+
+You can follow the 3.x guide at:
+
+  * [Linux installation](https://docs.espressif.com/projects/esp-idf/en/v3.3.1/get-started/linux-setup.html)
+  * [MacOS installation](https://docs.espressif.com/projects/esp-idf/en/v3.3.1/get-started/macos-setup.html)
+  * [Windows installation](https://docs.espressif.com/projects/esp-idf/en/v3.3.1/get-started/windows-setup.html)
 
-Once everything is set up you should have a functioning toolchain with
-prefix xtensa-esp32-elf- (or otherwise if you configured it differently)
-as well as a copy of the ESP-IDF repository. You will need to update your `PATH`
-environment variable to include the ESP32 toolchain. For example, you can issue
-the following commands on (at least) Linux:
+You will need to update your `PATH` environment variable to include the ESP32
+toolchain. For example, you can issue the following commands on (at least)
+Linux:
 
     $ export PATH=$PATH:$HOME/esp/crosstool-NG/builds/xtensa-esp32-elf/bin
 
-You can put this command in your `.profile` or `.bash_login`.
+You can put this command in your `.profile` or `.bash_login`, or do it manually.
+
+Configuring the MicroPython build
+---------------------------------
 
 You then need to set the `ESPIDF` environment/makefile variable to point to
-the root of the ESP-IDF repository.  You can set the variable in your PATH,
-or at the command line when calling make, or in your own custom `makefile`.
-The last option is recommended as it allows you to easily configure other
-variables for the build.  In that case, create a new file in the esp32
-directory called `makefile` and add the following lines to that file:
+the root of the ESP-IDF repository. The recommended way to do this is to have
+a custom `makefile` in `ports/esp32` which sets any additional variables, then
+includes the main `Makefile`. Note that GNU Make will preferentially run
+`GNUmakefile`, then `makefile`, then `Makefile`, which is what allows this to
+work. On case-insensitive filesystems, you'll need to use `GNUmakefile` rather
+than `makefile`.
+
+Create a new file in the esp32 directory called `makefile` (or `GNUmakefile`)
+and add the following lines to that file:
+
 ```
 ESPIDF = <path to root of esp-idf repository>
 BOARD = GENERIC
```

Comment:
Wouldn't it be better to use `?=` so the variables can be overridden on the commandline?

Reply:
The point of the custom makefile is so that you don't have to override each time. But if you want to also override the custom makefile that overrides the main makefile, sure, add some `?=`

Reply:
The latter is what I mean. At least I do it all the time, e.g., have GENERIC as default but build for TINYPICO when Seon has an issue...